analyzer/src/main/kotlin/managers/Yarn2.kt

```diff
@@ -126,21 +181,500 @@ class Yarn2(
+            if (dependencyPkg == null) {
+                log.warn { "Could not find package for sub dependency '$dependencyId' of package '$id'." }
+                null
+            } else {
+                val subDependencies = dependencyPkg.collectDependencies(
+                    allDependencies,
+                    ancestorPackageIds + dependencyId
+                )
+                YarnModuleInfo(dependencyId, dependencyPkg, subDependencies)
+            }
+        }?.toSet().orEmpty()
+    }
+
+    /**
+     * Process the [json] result of `yarn npm info` for a given package and return a populated [AdditionalData].
+     */
+    private fun processAdditionalPackageInfo(json: ObjectNode): AdditionalData {
+        val name = json["name"].textValue()
+        val version = json["version"].textValue()
+        val description = json["description"].textValueOrEmpty()
+        val vcsFromPackage = Npm.parseVcsInfo(json)
+        val homepage = json["homepage"].textValueOrEmpty()
+        val author = Npm.parseAuthors(json)
+
+        val dist = json["dist"]
+        var downloadUrl = dist["tarball"].textValueOrEmpty()
+
+        downloadUrl = Npm.fixDownloadUrl(downloadUrl)
+
+        val hash = Hash.create(dist["shasum"].textValueOrEmpty())
+
+        val vcsFromDownloadUrl = VcsHost.parseUrl(downloadUrl)
+
+        return AdditionalData(
+            name,
+            version,
+            description,
+            vcsFromPackage,
+            vcsFromDownloadUrl,
+            homepage,
+            downloadUrl,
+            hash,
+            author
+        )
+    }
+
+    /**
+     * Process [dependencies], the `Dependencies` sub-element of a single node returned by `yarn info`.
+     * The dependencies are returned as a list.
+     */
+    private fun processDependencies(dependencies: JsonNode): List<Identifier> =
+        dependencies.mapNotNull { dependency ->
+            val locator = dependency["locator"].textValue()
+            val locatorMatcher = requireNotNull(EXTRACT_FROMLOCATOR_PATTERN.matchEntire(locator)) {
+                "Locator '$locator' cannot be parsed."
+            }
+            val locatorRawName = locatorMatcher.groupValues[1]
+            val locatorType = locatorMatcher.groupValues[2]
+            val locatorVersion = locatorMatcher.groupValues[3]
+
+            val (locatorNamespace, locatorName) = Npm.splitNamespaceAndName(locatorRawName)
+            val version = cleanYarn2VersionString(locatorVersion)
+
+            val identifierType = if ("workspace" in locatorType) "Yarn2" else "NPM"
+            when {
+                // Prevent @virtual dependencies because they are internal to Yarn.
+                // See https://yarnpkg.com/features/protocols
+                "virtual" in locatorType -> null
+
+                else -> {
+                    kotlin.runCatching {
+                        Identifier(identifierType, locatorNamespace, locatorName, version)
```

Comment:
Question: Is this done in a `runCatching` block because the single identifier components could contain illegal characters, i.e. a ':' ?
Please remove the `kotlin.` prefix before `runCatching`.

Reply:
Thanks.